workshops/boost-research-reproducibility-binder/workshop-presentations/zero-to-binder-julia.md

```diff
@@ -0,0 +1,296 @@
+
+It is not practical to place large files in your GitHub repo or include them directly in the image that Binder builds.
+The best option for large files is to use a library specific to the data format to stream the data as you're using it or to download it on demand as part of your code.
+
+For security reasons, the outgoing traffic of your Binder is restricted to HTTP or GitHub connections only. You will not be able to use FTP sites to fetch data on mybinder.org.
+
+### Private files
+
+There is no way to access files which are not public from mybinder.org.
+You should consider all information in your Binder as public, meaning that:
+
+- there should be no passwords, tokens, keys etc in your GitHub repo;
+- you should not type passwords into a Binder running on mybinder.org;
+- you should not upload your private SSH key or API token to a running Binder.
+
+In order to support access to private files, you would need to create a local deployment of [BinderHub](https://binderhub.readthedocs.io/en/latest/) where you can decide the security trade-offs yourselves.
+
+**N.B.:** Building a BinderHub is not a simple task and is usually taken on by IT/RSE groups for reasons around managing maintenance, security and governance.
+However, that is not to say that they are the _only_ groups of people who should/could build a BinderHub.
+
+## 8. Get data with `postBuild`
+
+**TO DO:** :vertical_traffic_light:
+
+1) Go to your GitHub repo and create a file called `postBuild`
+2) In `postBuild`, add a single line reading: `wget -q -O gapminder.csv http://bit.ly/2uh4s3g`
+   - `wget` is a program which retrieves content from web servers. This line extracts the content from the bitly URL and saves it to the file denoted by the `-O` flag (i.e. `gapminder.csv`). The `-q` flag tells `wget` to do this quietly, i.e. don't print anything to the console.
+3) Update your `Project.toml` file by adding new dependencies to `[deps]` with the following lines:
+
+   ```julia
+   DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
+   Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+   ```
+
+   - These packages aren't necessary to download the data but we will use them to read the CSV file and make a plot
+4) Click the binder badge in your README to launch your Binder
+
+Once the Binder has launched, you should see a new file has appeared that was not part of your repo when you clicked the badge.
+
+Now visualise the data by creating a new notebook ("New" :arrow_right: "Julia") and run the following code in a cell.
+
+```julia
+using DataFrames
+using CSV
+using Plots
+
+data = CSV.read("gapminder.csv")
+
+# Transpose DataFrame
+numeric_tr = Matrix(data[:,2:end])'
+data_tr = DataFrame(numeric_tr, Symbol.(data[:, :country]))
+
+# Extract Year and add to DataFrame
+column_names = names(data)
+str_years = [String(x)[end-3:end] for x in column_names[2:end]]
+years = parse.(Int, str_years)
+data_tr[:, :Year] = years
+
+# Plot
+plot(data_tr[:, :Year], data_tr[:, :Australia])
+```
```

Comment:
I know this is my own code, but it was developed to mimic the actions of the Python variant. Since it doesn't use built-in `pandas` functions, it's much less clear why we might want to do this or indeed what it's doing. Even as a Julia user, I don't find it especially intuitive.

One might consider something simpler if the plot is the primary purpose of this:

```julia
using DataFrames
using CSV
using Plots

data = CSV.read("gapminder.csv")

# Extract the row corresponding to Australia
aus_gdp = data[data[:, :country] .== "Australia", :]
aus_gdp = Matrix(aus_gdp[:,2:end])[:]  # as vector

# Extract the years as Ints from the column names
years = [x[end-3:end] for x in names(data)[2:end]]
years = parse.(Int, years)

# Plot
plot(years, aus_gdp)
```

Reply:
Yes, thank you. The plot is the purpose of this section and a crazy dream of mine is that I could run a Zero to Binder tutorial and attendees could choose their preferred language within the same session. The point of the tutorial is to learn how to use Binder for Julia users, not get caught up with the workings of Julia. So I agree, let's go with the more intuitive example.